tools/profiler/profiler.py

```diff
@@ -0,0 +1,815 @@
+            stat_collector_in = get_raw_stat_collector(num_samples=subset_size)
+            statistic_container.add_statistic_point(
+                StatisticPoint(target_point=statistic_point_in, tensor_collector=stat_collector_in, algorithm="collect")
+            )
+
+        return statistic_container
+
+    def collect_activations(
+        self,
+        model: ov.Model,
+        dataset: Optional[Any] = None,
+        pattern: Optional[Union[str, Pattern[str]]] = None,
+        num_samples: Optional[int] = None,
+    ) -> ActivationData:
+        """
+        Collect activation statistics from layers matching the specified pattern.
+
+        This method profiles the model by running inference on the dataset and collecting
+        raw activation values at both the input and output of matching layers.
+
+        Args:
+            model: OpenVINO model to profile
+            dataset: Optional dataset to use for collection. If None, uses the instance's dataset.
+                    Should be compatible with NNCF's OVStatisticsAggregator
+            pattern: Optional regex pattern to match layer names. If None, uses the instance's pattern.
+                    Examples: r'self_attn', r'__module.model.layers.\d+.mlp'
+            num_samples: Optional number of samples to collect. If None, uses the instance's num_samples
+
+        Returns:
+            Dictionary mapping layer names to their collected activations:
+            {
+                'layer_name_1': {
+                    'in': [array1, array2, ...],   # List of input activation arrays
+                    'out': [array1, array2, ...]   # List of output activation arrays
+                },
+                'layer_name_2': {...},
+                ...
+            }
+            Each array is a numpy array containing activation values for one sample.
+
+        Raises:
+            ValueError: If no layers match the specified pattern
+        """
+        # Use provided parameters or fall back to instance defaults
+        dataset = dataset if dataset is not None else self.dataset
+        pattern = pattern if pattern is not None else self.pattern
+        num_samples = num_samples if num_samples is not None else self.num_samples
+
+        # Compile pattern if it's a string
+        regexp: Pattern[str] = re.compile(pattern) if isinstance(pattern, str) else pattern
+
+        # Build graph and create statistics aggregator
+        graph = GraphConverter.create_nncf_graph(model)
+        statistics_aggregator = OVStatisticsAggregator(dataset)
+
+        # Find target nodes matching the pattern
+        node_keys = graph.get_all_node_keys()
+        target_names = [key for key in node_keys if regexp.search(key)]
+
+        if not target_names:
+            msg = f"No layers found matching pattern: {pattern}"
+            raise ValueError(msg)
+
+        target_ops = [graph.get_node_by_key(name) for name in target_names]
```

Comment:
Sort nodes in topographical order. Currently nodes are sorted by lexicographical order
```suggestion
        target_ops = []
        for node in graph.topological_sort():
            if len(target_ops) == len(target_names):
                break
            if node.node_key in target_names:
                target_ops.append(node)
```